doc/developer/design/20210713_S3_sources_with_headers.md

```diff
@@ -0,0 +1,135 @@
+should be extendable with the current work, but will be prioritized separately.
+
+## Description
+
+<!--
+// Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+// If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+User-facing changes are limited to an expansion of the ability to combine
+existing formats with S3 sources, for example e.g. `FORMAT CSV WITH HEADER`.
+
+When a user specifies a source format that has a header, we will obtain an
+arbitrary file from the source declaration and inspect its header, making it the
+cannonical schema. All files must have exactly the same schema -- if any are
+encountered that do not match, the dataflow will be put into an error state.
+
+Internally, there are only a few changes that need to be made.
+
+* Add some code in purify that can obtain an object according to the source
+  declaration and use it to extract the schema.
+* Modify the messages sent from sources to include a signifier for "beginning of
+  object", which will cause the decode loop to perform schema validation and
+  drop the header line.
+
+### Purification strategy
+
+We will extend the [purification strategy from existing formats][csv-purify]
+into their combination with S3, including using existing syntax.
+
+[csv-purify]: https://github.com/MaterializeInc/materialize/blob/88cf93c3309ca62/src/sql/src/pure.rs#L480-L501
+
+#### Examples
+
+A CSV source containing files with header lines like `id,value` as:
+
+```sql
+CREATE SOURCE example
+FROM S3 DISCOVER OBJECTS USING BUCKET SCAN 'bucket'
+FORMAT CSV WITH HEADER;
+```
+
+will be rewritten via purify to:
+
+```sql
+CREATE SOURCE example (id, value)
+FROM S3 DISCOVER OBJECTS USING USING BUCKET SCAN 'bucket'
+FORMAT CSV WITH HEADER;
+```
+
+Conversely, the following create source statement (and the equivalent without
+headers) will be rejected:
+
+```sql
+CREATE SOURCE example
+FROM S3 DISCOVER OBJECTS USING SQS NOTIFICATIONS 'queuename'
+FORMAT CSV WITH HEADER;
+```
+
+it will instead require that users specify the column names:
+
+```sql
+CREATE SOURCE example (id, value)
+FROM S3 DISCOVER OBJECTS USING SQS NOTIFICATIONS 'queuename'
+FORMAT CSV WITH HEADER;
+```
+
+Similar rules will apply to Avro OCF objects, when implemented.
+
+## Alternatives and future work
+
+### Supporting a subset of schema fields
+
+For some use cases -- imagine an ETL job that has added or dropped columns over
+time -- it would be useful to support only a few required fields, or allowing a
+subset or reordering of fields.
+
+Supporting this is not planned for the initial implementation, but may be
+considered if it is requested.
+
+### Lazy sources
+
+For S3 sources, it's possible to specify that the *only* objects to ingest are
+the ones specified in an SQS queue. For this use case, reading from the SQS
+queue in order to determine what the header should be would be destructive.
```

Comment:
Not necessarily advocating for this, but reading from SQS isn't destructive unless you explicitly delete what you've read. You could always attempt to read one message and then terminate the visibility timeout: https://docs.amazonaws.cn/en_us/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html#changing-message-visibility-timeout.

Reply:
Unfortunately there is no guaranteed message delivery order, so if you don't delete messages then there is no guarantee of progress unless you delete messages, AFAIK. Certainly there's no guarantee of termination -- we'd just continuously get new messages from SQS in a hot loop if we don't delete them.

Reply:
No, I mean in `purify`, if you wanted to read just one message from the SQS topic and then put it back, that's doable.

Reply:
Oh, true! Yeah that has the same category of failure mode as scanning an empty bucket, so there is at least an argument that it should be equally valid to attempt.